### PR TITLE
New Version 2 codebase

### PR DESCRIPTION
**Pull request by cidrblock**
This is a complete rewrite of the plugin to support tox v4 and a scope currently very specific to the faciliation of collection testing.

Additional capabilities and PR are, of course, always welcome.

### CODE REVIEW

File: .github/workflows/run.yml
Comment (greg-hellings) [inline .github/workflows/run.yml:1]: Are we sure that we want this in the workflows for this repository? This repo isn't likely to host Ansible code itself, is it? I realize the point is to have it shared across other users of the plugin if they don't have a reason to customize the matrix, but it seems unnecessary to have it run on this repo directly.

File: README.md
Comment (greg-hellings) [inline README.md:162]: I would like to see the format of the `--gh-matrix` output validated more. Personally when I have code that runs on GHA and uses a container I run it in an `ubuntu-latest` and when it uses vagrant for the driver I run it on a `macos-latest` runner because that has Vagrant support.

I would like to be able to do the same in the future, so documentation of the output values here would permit me to craft my usage of it better.

File: src/ansitest/plugin.py
Comment (greg-hellings) [inline src/ansitest/plugin.py:42]: Is it the intention to frequently update this plugin to only support these versions of Ansible? Is the user able to customize the Ansible versions they want to test? (Also, these seem to be `ansible-core` versions and not full Ansible release versions?)

**Review by greg-hellings — CHANGES_REQUESTED** [review]

File: src/ansitest/plugin.py
Comment (cidrblock) [reply src/ansitest/plugin.py:495]: Funny I was thinking about this over the weekend, we should definately allow the user to provide some customization of the environment variables.

I was thinking they could go in the ansible section of the tox.ini so they would not interfere with other environments...

Comment (greg-hellings) [reply src/ansitest/plugin.py]: So it would probably be even better if this could be configured on a per-env basis. I wonder if, instead of injecting this MemoryLoader with `.insert(0, loader)` you could do `.append(loader)` and have the pre-programmed values in the plugin be the defaults for Ansible test cases while allowing them to be overridden in the tox.ini file?

Like, could I have

```ini
[testenv:unit-py3.11-2.14]
passenv = MYENV_*
```

and accept the rest of the values from this plugin but just override that? In my local testing that is exactly how it works. When I change the insert to an append, it allows the user to configure the test case however they want in the tox.ini but still accept my defaults from a plugin. Then we can have very conservative defaults, like this one, and allow the user to override it if they are running in Jenkins, CircleCI, etc.

Comment (cidrblock) [reply src/ansitest/plugin.py]: If I recall, the reason I went down the path of a new MemoryLoader was, if I recall, there was some attributes I couldn't modify. I can try it again.

For now, I've extracted the values from the ini loader and added them to the custom values in the MemoryLoader, it isn't great.

File: src/ansitest/plugin.py
Comment (cidrblock) [reply src/ansitest/plugin.py]: We're still trying to sort out molecule here, deciding if it should be here or tox should just leverage the pytest-ansible project which will/is getting molecule support..

Comment (cidrblock) [reply src/ansitest/plugin.py]: We're still trying to sort out molecule here, deciding if it should be here or tox should just leverage the pytest-ansible project which will/is getting molecule support..

File: .github/workflows/run.yml
Comment (cidrblock) [reply .github/workflows/run.yml:1]: It should be disabled now in the actions, yeah, the intention wasn't to have it run here, just hosted.

I'll talk to sorin about this, it might make sense to have them all in one palce outside of the py project repos.

File: README.md
Comment (cidrblock) [reply README.md:162]: I added a test for the GH json, I'll try to remember to add a README entry for it tomorrow :)

File: src/ansitest/plugin.py
Comment (cidrblock) [reply src/ansitest/plugin.py]: Yes, the goal is for this to be maintained.  If we can get some adaption across any of the community collection maintainers or internal ansible content engineering teams, I don't think this should add much operational overhead.

Yes, the goal here now was for collection testing which shouldn't require the full ansible package.  Any collection should have it's deps listed in it's requirements file, which is part of the reason the collection is copied to the tox venv, built, and installed.  the act of installation pulls in those other collection deps.

**Comment by cidrblock** [comment]
@greg-hellings TY for taking the time to review....  I'm still sorting out the tox internals, so I suspect there are better ways to do some of this. LMK

File: src/tox_ansible/plugin.py
Comment (greg-hellings) [inline src/tox_ansible/plugin.py:204]: If you change THIS line, you can skip all the stuff above and just leave it alone. Then the user will be able to override the dfeaults in this plugin by manually entering them into the ini file. This is because the internals of tox iterate through the `env_conf.loaders` and query each entry for a value, then take the first value one of the loaders returns. When you have the line as currently written, then the plugin takes precedence over anything the user sets as a value. When you change it to my suggestion, the value in the ini file takes precedence, and the values of this plugin fill in default values.

You should revert the code above this back to its simple form of just providing the default values, then use the suggestion I'm putting here to allow INI file overrides.

```suggestion
    env_conf.loaders.append(loader)
```

Comment (greg-hellings) [reply src/tox_ansible/plugin.py:204]: Given this sample INI file entry

```ini
[testenv:a]
commands = echo hi from tox.ini
```

And the following plugin code:

```python
env_conf.loaders.insert(0, MemoryLoader(commands=["echo hi"], deps=["ansible"]))
```

I get the following by running `tox -e a`:
`a: commands[0]> echo hi`

If I change the Python line to read
```python
env_conf.loaders.append(MemoryLoader(commands=["echo hi"], deps=["ansible"]))
```

I get the following by running `tox -e a`:
`a: commands[0]> echo hi from tox.ini`

Both of them still have Ansible in the list of deps, because I do not explicitly set any value for deps in the INI file. However, I am able to customize any portion of it I want, using standard Tox methods in the tox.ini file. Personally, I think this would be superior to appending to them or to overriding them with the previous code.

File: src/ansitest/plugin.py
Comment (greg-hellings) [reply src/ansitest/plugin.py]: No, that's not quite what I mean. I'll comment on the line that I mean, along with the examples I ran to validate what Tox is doing internally.

Comment (cidrblock) [reply src/ansitest/plugin.py]: Got it, I mis understood, agreed, let it work like native tox.  `append` is much cleaner

Comment (cidrblock) [reply src/ansitest/plugin.py]: I did find one thing interesting switching to append from insert. Any configuration in `[testenv]` overrides the configuration injectect by the plugin. I would have expected the environment specific configuration provided by the plugin to override anything in the root ini configuration.

I added some explanation to the README

Comment (greg-hellings) [reply src/ansitest/plugin.py]: Hmmm, if you think that's undesirable we could maybe have it insert to the -1 position of the list? But then you're really relying on specific internals of tox.ini to have its loaders all in a specific order. In my opinion, the behavior you mentioned is fine, as it is still explicit configuration in `[testenv]` overriding implicit configuration. How do you feel on it?

Comment (cidrblock) [reply src/ansitest/plugin.py]: I think it's fine, although unexpected.  If someone want to modify the tox-ansible provided configuration, they have a few choices, redefine the provided in the custom, raise a PR/issue to modifu the default, or not use the plugin at all.

The only thing that leaves me a little worried is that I had envisioned this being a plugin that would work along side any exisitng tox.ini file without modification, and be enabled with the `--ansible` flag. Thinking about it now, the use could specific a different tox.ini file for ansible related activity or we could possibly inject the alternate for them (basically change the default from tox.ini to tox-ansible.ini.  I'll do some testing here.

Comment (cidrblock) [reply src/ansitest/plugin.py]: The plugin now suggests the use of `tox-ansible.ini` instead of `tox.ini` and will warn if the user is using the later. I think it is cleaner (and safer) and allows the plugin to be adopted without modifications of an existing ini file that might be in place.

LMK what you think

File: src/ansitest/plugin.py
Comment (cidrblock) [reply src/ansitest/plugin.py]: @greg-hellings We are going to add molecule support back in after version 2 is released, there's a couple reason here, molecule 6 is on the horizon, there is work to merge pytest-molecule into pytest-ansible.

One thing we are talking about is having tox-ansible run a pytest -coolect-only and using that to dynamically build the tox environment list based on a one test-prescenario, and then add in the python/core version specified by the user.

File: .github/workflows/run.yml
Comment (cidrblock) [reply .github/workflows/run.yml:1]: I think we can just punt on this for now........  We are talking about we can centalize these and where it's appropriate.  The other thing that comes into play is the idea of a "content developers kit" for ansible that acts as an installer for a number of python project scattered all over the place. 

So for now, I say we leave the workflow here until we solve the bigger structure question for all the dev tools.

**Comment by cidrblock** [comment]
@ssbarnea @greg-hellings I think this is ready.  We can sort out the release pipeline next week once this is merged and put a 2.0 out on pypi...

File: .flake8
Comment (greg-hellings) [inline .flake8:9]: Should this also include `.tox` so it doesn't try to flake8 the entire test folder structure?

**Review by greg-hellings — APPROVED** [review]

File: .flake8
Comment (cidrblock) [reply .flake8:9]: yup, good catch, TY

**Review by greg-hellings — APPROVED** [review]

File: .github/workflows/tox.yml
Comment (cidrblock) [inline .github/workflows/tox.yml:28]: err, 3.9+

**Pull request merged**
